lib/modules/manager/bazel-module/fragments.ts

```diff
@@ -0,0 +1,294 @@
+  addValue(item: ValueFragment): void {
+    this.items.push(item);
+  }
+}
+
+export class RecordFragment {
+  static as(data: unknown): RecordFragment {
+    return as(
+      data,
+      'record',
+      RecordFragment,
+      (v: ValueFragmentInterface): v is RecordFragmentInterface =>
+        v.type === 'record',
+      RecordFragment.from
+    );
+  }
+
+  static from(frag: RecordFragmentInterface): RecordFragment {
+    const children: ChildFragments = Object.entries(frag.children).reduce(
+      (acc, propVal) => {
+        acc[propVal[0]] = asValue(propVal[1]);
+        return acc;
+      },
+      {} as ChildFragments
+    );
+    return new RecordFragment(children, frag.isComplete);
+  }
+
+  static safeAs(data: unknown): RecordFragment | null {
+    try {
+      return RecordFragment.as(data);
+    } catch (e) {
+      return null;
+    }
+  }
+
+  readonly type: FragmentType = 'record';
+  isComplete: boolean;
+  children: ChildFragments;
+
+  constructor(children: ChildFragments = {}, isComplete = false) {
+    this.children = children;
+    this.isComplete = isComplete;
+  }
+
+  addAttribute(attrib: AttributeFragment): void {
+    if (!attrib.value) {
+      throw new Error('The attribute fragment does not have a value.');
+    }
+    this.children[attrib.name] = attrib.value;
+  }
+}
+
+export class AttributeFragment {
+  static as(data: unknown): AttributeFragment {
+    if (data instanceof AttributeFragment) {
+      return data;
+    }
+    const frag = AttributeFragmentSchema.parse(data);
+    return new AttributeFragment(
+      frag.name,
+      frag.value ? asValue(frag.value) : undefined
+    );
+  }
+
+  static safeAs(data: unknown): AttributeFragment | null {
+    try {
+      return AttributeFragment.as(data);
+    } catch (e) {
+      return null;
+    }
+  }
+
+  readonly type: FragmentType = 'attribute';
+  readonly name: string;
+  value?: ValueFragment;
+
+  constructor(name: string, value?: ValueFragment) {
+    this.name = name;
+    this.value = value;
+  }
+
+  get isComplete(): boolean {
+    return is.truthy(this.value);
+  }
+
+  addValue(item: ValueFragment): void {
+    this.value = item;
+  }
+}
+
+export type ValueFragment =
+  | ArrayFragment
+  | BooleanFragment
+  | RecordFragment
+  | StringFragment;
+export type ChildFragments = Record<string, ValueFragment>;
+export type Fragment = ValueFragment | AttributeFragment;
+
+// Static Functions
+
+function as<I extends ValueFragmentInterface, F>(
+  data: unknown,
+  expected: string,
+  fragCtor: new (...args: any[]) => F,
+  iTypeGuard: (v: ValueFragmentInterface) => v is I,
+  fromFn: (frag: I) => F
+): F {
+  if (data instanceof fragCtor) {
+    return data;
+  }
+  const fragInterface = ValueFragmentSchema.parse(data);
+  if (iTypeGuard(fragInterface)) {
+    return fromFn(fragInterface);
+  }
+  throw new Error(
+    'The data is a value fragment, but it is not the expected type.' +
+      `expected: ${expected}, actual: ${fragInterface.type}.`
```

Comment:
This error is way better, thank you for improving it! ❤️ 

Is this supposed to be one or two sentences? If it's two we should capitalize the first letter of the second sentence:

```suggestion
    'The data is a value fragment, but it is not the expected type.' +
      `Expected: ${expected}, actual: ${fragInterface.type}.`
```